AwsCryptographicMaterialProviders/dafny/AwsCryptographicMaterialProviders/src/CMCs/StormTracker.dfy

```diff
@@ -39,11 +40,65 @@ module {:options "/functionSyntax:4" }  StormTracker {
       gracePeriod := 10,
       graceInterval := 1,
       fanOut := 20,
-      inFlightTTL := 20,
+      inFlightTTL := 10,
       sleepMilli := 20
     )
   }
 
+  predicate ConsistentSettings(
+    cache: Types.StormTrackingCache
+  ) {
+    && cache.graceInterval <= cache.gracePeriod
+    && cache.inFlightTTL <= cache.gracePeriod
+    && cache.graceInterval <= cache.inFlightTTL
+  }
+
+  function N(n : Types.CountingNumber) : string {
+    String.Base10Int2String(n as int)
+  }
+
+  function BadCacheMsg(cache: Types.StormTrackingCache) : string
+    requires !ConsistentSettings(cache)
+  {
+    var msg := "For a StormCache : ";
+
+    var msg := msg +
+
+               if !(cache.graceInterval <= cache.gracePeriod) then
+                 "graceInterval must not exceed gracePeriod, yet configuration has graceInterval="
+                 + N(cache.graceInterval) + " and gracePeriod=" + N(cache.gracePeriod) + ". "
+               else
+                 "";
+
+    var msg := msg +
+               if !(cache.inFlightTTL <= cache.gracePeriod) then
+                 "inFlightTTL must not exceed gracePeriod, yet configuration has inFlightTTL="
+                 + N(cache.inFlightTTL) + " and gracePeriod=" + N(cache.gracePeriod) + ". "
+               else
+                 "";
+
+    var msg := msg +
+               if !(cache.graceInterval <= cache.inFlightTTL) then
+                 "graceInterval must not exceed inFlightTTL, yet configuration has graceInterval="
+                 + N(cache.graceInterval) + " and inFlightTTL=" + N(cache.inFlightTTL) + ". "
+               else
+                 "";
+
+    msg
+  }
+
+  function CheckSettings(
+    cache: Types.StormTrackingCache
+  ) : (res : Result<bool, Types.Error> )
+    ensures res.Success? ==> ConsistentSettings(cache)
+  {
+    if ConsistentSettings(cache) then
+      Success(true)
+    else
+      Failure(Types.AwsCryptographicMaterialProvidersException(message := BadCacheMsg(cache)))
+  }
```

Comment:
What about this?  I also feel like we could use a `Need` that runs the error message in a thunk, what do you think?

```suggestion
  function CheckSettings(
    cache: Types.StormTrackingCache
  ) : (res : Outcome<Types.Error> )
    ensures res.Pass? ==> ConsistentSettings(cache)
  {
    if ConsistentSettings(cache) then
      Pass
    else
      Fail(Types.AwsCryptographicMaterialProvidersException(message := BadCacheMsg(cache)))
  }
```